Source/WebKit/UIProcess/WebPageProxy.cpp

```diff
@@ -8840,6 +8840,12 @@ void WebPageProxy::revokeGeolocationAuthorizationToken(const String& authorizati
     m_geolocationPermissionRequestManager.revokeAuthorizationToken(authorizationToken);
 }
 
+static bool isOriginUnique(const SecurityOriginData& origin)
```

Comment:
Like I mentioned in my previous review, I think we should do the isUnique() check in WebCore's query(), on a SecurityOrigin. 

Relying on data members of a SecurityOriginData to be empty (but not null) to determine that the origin is unique is a bit obscure and fragile.

Reply:
The UIProcess side check can then be a MESSAGE_CHECK and this function can be name something like isValidOrigin().

Reply:
Besides clarity, another reason for my proposal is that there is no good reason to do IPC or involve the UIProcess for unique/opaque origins. It is more efficient to deal with them in the WebProcess.

Reply:
I think it is best to return 'denied' if possible (say application does not have the camera entitlement), or reject with NotSupported (say permission name is background-fetch).
This is best done in UIProcess.

Reply:
I don't understand what that means, you are returning Prompt unconditionally so...

Reply:
Also, whatever decisions about entitlements or not supporting a particular API should be doable at WebProcess-level if we wanted to.

Reply:
We also have plenty of checks in the WebProcess that resolve/reject the promise early in WebCore's query().

Reply:
Also, we normally don't pass opaque origins to other processes since they are only valid within a single process (requires pointer identity). This is why we have a concept of unique/opaque SecurityOrigin but we don't have one for SecurityOriginData. Relying of SecurityOriginData having non-null / empty data members to make such determination is not robust or clear.